pydantic/types.py

```diff
@@ -237,4 +243,36 @@ class NegativeFloat(ConstrainedFloat):
     lt = 0
 
 
+class ConstrainedUUID(UUID):
```

Comment:
`ConstrainedUUID` could lead to confusion, I think we can remove it altogether.

I think best to just have 

```py
class UUID1(UUID):
    _required_version = 1

...
```

Then change `uuid_validator`:

```py
def uuid_validator(v) -> UUID:
    if isinstance(v, str):
        v = UUID(v)
    elif isinstance(v, (bytes, bytearray)):
        v = UUID(v.decode())
    elif not isinstance(v, UUID):
        raise ValueError(f'str, byte or native UUID type expected not {type(v)}')
    required_version = getattr(v, '_required_version', None)
    if required_version and required_version != v.version:
        ...
    return v
```

Reply:
This is a good idea, but I think this is impossible to implement (or I don't know how :). As I see in my tests in `uuid_validator` `v` is instance of `UUID` type, not `UUID1` or `UUID2`... 

Validator implementation:
```
def uuid_validator(v) -> UUID:
    if isinstance(v, str):
        v = UUID(v)
    elif isinstance(v, (bytes, bytearray)):
        v = UUID(v.decode())
    elif not isinstance(v, UUID):
        raise ValueError(f'str, byte or native UUID type expected not {type(v)}')

    required_version = getattr(v, '_required_version', None)
    if required_version and v.version != required_version:
        raise ValueError(f'uuid version {required_version} expected, not {v.version}')

    return v
```

And test:
```
class UUIDModel(BaseModel):
    a: UUID1
    b: UUID3
    c: UUID4
    d: UUID5


def test_uuid_validation():
    a = uuid.uuid1()
    b = uuid.uuid3(uuid.NAMESPACE_DNS, 'python.org')
    c = uuid.uuid4()
    d = uuid.uuid5(uuid.NAMESPACE_DNS, 'python.org')

    m = UUIDModel(a=a, b=b, c=c, d=d)
    assert m.dict() == {
        'a': a,
        'b': b,
        'c': c,
        'd': d,
    }

    with pytest.raises(ValidationError) as exc_info:
        UUIDModel(a=d, b=c, c=b, d=a)
    assert exc_info.value.message == '4 errors validating input'
```

Reply:
Done! :)